wrappers/python/openmm/app/simulation.py

```diff
@@ -252,23 +272,43 @@ def _simulate(self, endStep=None, endTime=None):
                     self._generate_reports(unwrapped, False)
     
     def _generate_reports(self, reports, periodic):
-        getPositions = False
-        getVelocities = False
-        getForces = False
-        getEnergy = False
-        for reporter, next in reports:
-            if next[1]:
-                getPositions = True
-            if next[2]:
-                getVelocities = True
-            if next[3]:
-                getForces = True
-            if next[4]:
-                getEnergy = True
-        state = self.context.getState(positions=getPositions, velocities=getVelocities, forces=getForces,
-                                      energy=getEnergy, parameters=True, enforcePeriodicBox=periodic,
+        '''
+        reports is a tuple of (reporter, nextReport)
+        reporter is the corresponding reporter object
+        nextReport is a description on what the reporter needs in the context. This can either be a dict, or a tuple (old).
+        '''
+        # Separate into new and old format for reports
+        oldReports = []
+        newReports = []
+        for report in reports:
+            if isinstance(report[1], dict):
+                newReports.append(report)
+            else:
+                oldReports.append(report)
+
+        # Check if any reporter needs positions, velocities,...
+        # will return False if newReports is empty
+        positions = any(['positions' in report[1]['include'] for report in newReports])
+        velocities = any(['velocities' in report[1]['include'] for report in newReports])
+        forces = any(['forces' in report[1]['include'] for report in newReports])
+        energy = any(['energy' in report[1]['include'] for report in newReports])
+        integratorParameters = any(['integratorParameters' in report[1]['include'] for report in newReports])
+        parameterDerivatives = any(['parameterDerivatives' in report[1]['include'] for report in newReports])
```

Comment:
I suggest doing this a little differently.

```python
includes = set.union(*[set(report['include']) for report in reports])
includeArgs = {include:True for include in includes}
```

And then in the call to `getState()`, pass `**includeArgs` as an argument list.  Advantages of this are 1) it's less code, 2) it's forward compatible if we add other options to `getState()` in the future, and 3) if a reporter returns an invalid option in `include`, the error will be caught and produce an exception.

Reply:
Good suggestion, this is certainly more readable. I made a slight modification to use a set comprehension and it now reads
`includes = {property for report in reports for property in report[1].get('include', [])}`

Reply:
I'm staring at that comprehension and totally failing to comprehend it. :)  It looks to me as if you're using `report` before defining it?  Perhaps it should be `{property for property in (report[1].get('include', []) for report in reports)}`?

Reply:
Nested list comprehensions need some getting used to

```python
includes = set()
for report in reports:
   for property in report[1].get('include', []):
      includes.add(property)
```

We start with the last expression, but then we need to go from top to bottom:

`{property for report in reports for property in report[1].get('include', [])}`

See [here](https://stackoverflow.com/questions/18072759/how-can-i-use-list-comprehensions-to-process-a-nested-list) for reference.

Reply:
Let's write it in a way that doesn't need getting used to. :)  Making code clear and easy to understand is important!  An advantage of `union()` is that it tells you at a glance what the code is doing: taking the union of the sets.